### PR TITLE
Web3d quantized attributes

### PR DESCRIPTION
**Pull request by lasalvavida**
Adds support for quantized attributes.

Related to [#308](https://github.com/AnalyticalGraphicsInc/agi-asset-pipeline/pull/308)

### CODE REVIEW

**Comment by lasalvavida** [comment]
@pjcozzi

**Comment by pjcozzi** [comment]
I'll do a first review here since I have also been looking at https://github.com/AnalyticalGraphicsInc/cesium/pull/3891, but @lilleyse can you please do the final review and merge?

**Comment by pjcozzi** [comment]
@JoshuaStorm @likangning93 can you make a new branch and include this in your interface?

File: lib/getBinaryGltf.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+var defined = Cesium.defined;
+var defaultValue = Cesium.defaultValue;
+var sizeOf = require('image-size');
+var mime = require('mime');
+var mergeBuffers = require('./mergeBuffers');
+var removePipelineExtras = require('./removePipelineExtras');
+
+module.exports = getBinaryGltf;
+
+function getBinaryGltf(gltf, callback) {
```
Comment (pjcozzi) [inline lib/getBinaryGltf.js]: Are there unit tests for this?

Comment (lasalvavida) [inline lib/getBinaryGltf.js]: There are unit tests for writeBinaryGltf which uses this directly.

File: lib/getBinaryGltf.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    gltf.buffers = defaultValue(gltf.buffers, {});
+    mergeBuffers(gltf, 'binary_glTF');
+
+    var bufferViews = gltf.bufferViews;
+    var buffers = gltf.buffers;
+    var currentOffset = buffers.binary_glTF.byteLength;
+    var body = buffers.binary_glTF.extras._pipeline.source;
+    var currentBinaryView = 0;
+    //Update object with KHR_binary_glTF properties and add to body and bufferViews
+    function updateBinaryObject(name) {
```
Comment (pjcozzi) [inline lib/getBinaryGltf.js]: I think it is better to put this function outside of the outer `getBinaryGltf` function and just pass in any data needs.  I expect it will get optimized better.

File: lib/getBinaryGltf.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        var objects = gltf[name];
+        if (defined(objects)) {
+            for (var objectId in objects) {
+                if (objects.hasOwnProperty(objectId)) {
+                    var object = objects[objectId];
+
+                    //Update object with binary format
+                    object.uri = "data:,";
```
Comment (pjcozzi) [inline lib/getBinaryGltf.js]: Use `'`.

File: lib/getBinaryGltf.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                    var bufferViewKeys = Object.keys(bufferViews);
+                    while (bufferViewKeys.indexOf('binary_bufferView' + currentBinaryView) != -1) {
+                        currentBinaryView++;
+                    }
+
+                    var objectSource = object.extras._pipeline.source;
+                    var bufferViewId = 'binary_bufferView' + currentBinaryView;
+                    KHR_binary_glTF.bufferView = bufferViewId; //Create bufferview
+                    bufferViews[bufferViewId] = {
+                        "buffer": "binary_glTF",
```
Comment (pjcozzi) [inline lib/getBinaryGltf.js]: Drop `"` for properties.  Use `'` for the string values.

I don't think gltf-pipeline is completely consistent here, but this is the standard Cesium.  There's some exceptions, for example, if the entire object is JSON (we do this for CZML examples in Sandcastle).

File: lib/getBinaryGltf.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    header.writeUInt32LE(glbLength, 8);
+    header.writeUInt32LE(sceneLength, 12);
+    header.writeUInt32LE(0, 16);
+
+    //Create scene buffer and overall buffer
+    var scene = new Buffer(sceneString);
+    var glb = Buffer.concat([header, scene, body], glbLength);
+    
+    if (callback) {
+        callback(header, scene, body);
```
Comment (pjcozzi) [inline lib/getBinaryGltf.js]: @leerichard42 shouldn't we use `process.nextTick()` throughout for callbacks?

File: lib/quantizeAttributes.js [hunk 6]
```diff
@@ -0,0 +1,348 @@
+'use strict';
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Did you check code coverage for this file?  See https://coveralls.io/github/AnalyticalGraphicsInc/gltf-pipeline

File: lib/getBinaryGltf.js (hunk 1 above)
Comment (pjcozzi) [inline lib/getBinaryGltf.js]: Sounds good, can you double check the coverage and update if needed?

File: lib/quantizeAttributes.js [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+ *                     "attributes": {
+ *                         "NORMAL": "accessor_0",
+ *                         "POSITION": "accessor_1"
+ *                     }
+ *                 }
+ *             ]
+ *         }
+ *     }
+ * };
+ * var options = {
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: What is the full documentation for `options`?

File: lib/quantizeAttributes.js [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+    var meshId;
+    var primitives;
+    var primitive;
+    var i, j;
+    var index;
+
+    // Retrieve the accessors that should be quantized
+    var quantizeAll = true;
+    if (defined(options)) {
+        attributes = options.attributes;
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Also move this into its own function.  I think it will make this function flow really well.

File: lib/quantizeAttributes.js [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+            // Quantize the data
+            var data = accessorData.data;
+            var encoded = new Uint16Array(data.length);
+            var index;
+            for (j = 0; j < data.length; j++) {
+                index = j % min.length;
+                encoded[j] = (data[j] - min[index]) * precision / (max[index] - min[index]);
+            }
+
+            //TODO: Repack the buffers once, don't do it for each accessor
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: What's the plan with this `TODO`?  We generally don't merge code with `TODO`s.

File: lib/quantizeAttributes.js [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+
+            // Quantize the data
+            var data = accessorData.data;
+            var encoded = new Uint16Array(data.length);
+            var index;
+            for (j = 0; j < data.length; j++) {
+                index = j % min.length;
+                encoded[j] = (data[j] - min[index]) * precision / (max[index] - min[index]);
+            }
+
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Extra whitespace.

File: lib/quantizeAttributes.js [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+                "WEB3D_quantized_attributes": {}
+            };
+            attributes = accessor.extensions.WEB3D_quantized_attributes;
+            var accessorData = getAccessorData(gltf, accessorId);
+            var min = accessorData.min;
+            var max = accessorData.max;
+
+            var decodeMatrix = [];
+            var precision = Math.pow(2, 16) - 1;
+            for (j = 0; j < min.length + 1; j++) {
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Is there a more straightforward - even if more verbose - way to create `decodeMatrix`?

File: lib/quantizeAttributes.js [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+                    else if (j == k) {
+                        decodeMatrix.push((max[j] - min[j]) / precision);
+                    }
+                    else {
+                        decodeMatrix.push(0);
+                    }
+                }
+            }
+            // decodedMin and decodedMax are required for POSITION attributes
+            if (attribute === "POSITION") {
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: `'`

File: lib/quantizeAttributes.js [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+                attributes.decodedMax = max;
+            }
+            attributes.decodeMatrix = decodeMatrix;
+
+
+            // Quantize the data
+            var data = accessorData.data;
+            var encoded = new Uint16Array(data.length);
+            var index;
+            for (j = 0; j < data.length; j++) {
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Throughout and in general, assign `data.length` to a local before the loop unless the length can change during the loop.

File: lib/quantizeAttributes.js [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+                }
+            }
+        }
+    }
+    // Add WEB3D_quantized_attributes extension to extensionsUsed
+    if (!defined(extensionsUsed)) {
+        extensionsUsed = [];
+        gltf.extensionsUsed = extensionsUsed;
+    }
+    extensionsUsed.push('WEB3D_quantized_attributes');
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Perhaps only add this if it doesn't already exist (make a helper function like `addExtensionsUsed(gltf, 'WEB3D_quantized_attributes')` that does all the required checks.

File: lib/quantizeAttributes.js [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+    data.max = [];
+
+    for (var i = 0; i < typedArray.length; i++) {
+        var element = typedArray[i];
+        var index = i % numComponents;
+        if (index >= data.min.length) {
+            data.min.push(element);
+            data.max.push(element);
+        }
+        else {
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Here and throughout if needed, put `else` on the same line as `}`.

File: lib/quantizeAttributes.js [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+            }
+            if (element > data.max[index]) {
+                data.max[index] = element;
+            }
+        }
+    }
+    return data;
+}
+
+function getByteLengthForComponentType(type) {
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Move this to its own file incase it is useful elsewhere.

Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Actually I could go either way.  Up to you.

File: lib/quantizeAttributes.js (hunk 14 above)
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: In general, does this stage workaround if some accessors are already quantized?

**Comment by pjcozzi** [comment]
Just those comments from me.

This might also need more unit tests.

**Comment by lasalvavida** [comment]
> This might also need more unit tests.

Yes, it definitely needs more robust testing

File: lib/writeBinaryGltf.js [hunk 17]
```diff
@@ -2,15 +2,7 @@
...
-var Cesium = require('cesium');
-var defined = Cesium.defined;
-var defaultValue = Cesium.defaultValue;
-var sizeOf = require('image-size');
-var mime = require('mime');
-var objectValues = require('object-values');
-var writeSource = require('./writeSource');
-var mergeBuffers = require('./mergeBuffers');
-var removePipelineExtras = require('./removePipelineExtras');
+var getBinaryGltf = requilre('./getBinaryGltf');
```
Comment (mlimper) [inline lib/writeBinaryGltf.js]: typo?

File: lib/quantizeAttributes.js [hunk 18]
```diff
@@ -0,0 +1,348 @@
...
+        case 5126:
+            return 4;
+    }
+}
+
+function arrayForComponentType(type, buffer, offset, length) {
+    var arrayBuffer;
+    switch (type) {
+        case 5120:
+            return new Int8Array(arrayBuffer);
```
Comment (mlimper) [inline lib/quantizeAttributes.js]: why is arrayBuffer undefined and still being used? is this intended?

File: lib/quantizeAttributes.js [hunk 19]
```diff
@@ -0,0 +1,348 @@
...
+            var encoded = new Uint16Array(data.length);
+            var index;
+            for (j = 0; j < data.length; j++) {
+                index = j % min.length;
+                encoded[j] = (data[j] - min[index]) * precision / (max[index] - min[index]);
+            }
+
+            //TODO: Repack the buffers once, don't do it for each accessor
+            //Adjust buffer to be float aligned
+            var shift = (encoded.length * 2) % 4;
```
Comment (mlimper) [inline lib/quantizeAttributes.js]: does this mean that the length of the quantized position array, for example, will be padded to the next 4-byte-boundary if necessary (using two additional zero bytes)?

File: lib/quantizeAttributes.js (hunk 7 above)
Comment (lasalvavida) [inline lib/quantizeAttributes.js]: `options` allows you to specify which primitives you want quantized. I'll make the documentation more robust. It's really just for generating test models though, in practice I think you would probably want to quantize everything possible, which is the default behavior if options is undefined.

File: lib/quantizeAttributes.js (hunk 9 above)
Comment (lasalvavida) [inline lib/quantizeAttributes.js]: There's nothing explicitly wrong with doing it the way it is, but it would probably be more efficient to reduce the amount of buffer modification. I'll try to get this fixed now.

File: lib/quantizeAttributes.js (hunk 11 above)
Comment (lasalvavida) [inline lib/quantizeAttributes.js]: Agrred, I'll see if I can find a way to make this more readable

File: lib/quantizeAttributes.js (hunk 14 above)
Comment (lasalvavida) [inline lib/quantizeAttributes.js]: It does not, but it definitely should.

File: lib/writeBinaryGltf.js (hunk 17 above)
Comment (lasalvavida) [inline lib/writeBinaryGltf.js]: Whoops, yes that's a typo

File: lib/quantizeAttributes.js (hunk 18 above)
Comment (lasalvavida) [inline lib/quantizeAttributes.js]: Nope, it can be removed, good catch

File: lib/quantizeAttributes.js (hunk 19 above)
Comment (lasalvavida) [inline lib/quantizeAttributes.js]: WebGL throws exceptions if data doesn't start on a multiple of its boundary size. We could look for unused bytes in the buffer and try to repack it to remove those holes, but I think that might be better as a separate stage.

**Comment by pjcozzi** [comment]
@mlimper thanks for the prompt review here!  We really value your support.

@lasalvavida sounds like you are making quick work of this.  Let me know when this is ready for another review.

File: lib/quantizeAttributes.js (hunk 19 above)
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: #50 should do this.

Comment (lasalvavida) [inline lib/quantizeAttributes.js]: It should do this, but I'm getting garbled models as in #57. I'll be pushing what I have very shortly, and then I'm going to take a look at fixing removeUnusedVertices.

Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Sounds good, thanks for looking at `removeUnusedVertices`.

**Comment by lasalvavida** [comment]
@pjcozzi I think I hit everything in the main function bodies. I'm working on tests and fixing removeUnusedVertices right now, but feel free to take a look at what's here

File: lib/getAccessorByteStride.js [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+ * If the byteStride is 0, it is calculated based on type and componentType
+ *
+ * @param {Object} [accessor] The accessor.
+ * @returns {Number} The byte stride of the accessor.
+ */
+function getAccessorByteStride(accessor) {
+    if (accessor.byteStride > 0) {
+        return accessor.byteStride;
+    }
+    else {
```
Comment (pjcozzi) [inline lib/getAccessorByteStride.js]: Drop the `else`.

See "Do not use an unnecessary else block at the end of a function" here: https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/CodingGuide#functions

File: lib/quantizeAttributes.js [hunk 21]
```diff
@@ -0,0 +1,281 @@
...
+
+/**
+ * Quantizes attributes in the provided glTF using the WEB3D_quantized_attributes extension.
+ * If options is undefined, all quantizable attributes will be quantized with the default options.
+ * If options.attributes is undefined, all quantizable attributes will be quantized with the provided options.
+ * If options.precision is undefined, the decodeMatrix be written in its full form.
+ *
+ * @param {Object} [gltf] A javascript object holding a glTF hierarchy.
+ * @param {Object} [options=undefined] Defines more specific quantization behavior.
+ * @param {Object} [options.attributes=undefined] Defines specific attributes to be quantized. This should be as arranged as meshId -> primitiveIndex -> [semantic_0, semantic_1, ... semantic_N]
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Do you think it needs to be this precise?  Imagine what a UI for this would look like.

Perhaps just start by providing an array of semantics, e.g., "I want to quantize all POSITIONS."

File: lib/quantizeAttributes.js [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+ *     }
+ * });
+ */
+function quantizeAttributes(gltf, options) {
+    var accessors = gltf.accessors;
+    var bufferViews = gltf.bufferViews;
+    var buffers = gltf.buffers;
+    var i, j;
+    var accessorIds = {};
+    var precision = undefined;
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Just:
```javascript
var precision;
```
`undefined` is implicit and I believe generates a JSHint or WebStorm warning.

File: lib/quantizeAttributes.js (hunk 21 above)
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: This would probably simplify the implementation a lot too.

Comment (lasalvavida) [inline lib/quantizeAttributes.js]: I needed it to be that precise for generating my test models. Particularly the mismatch models wouldn't be possible without it. That being said, I do think it would be a good idea to have another option for specifying semantics.

File: lib/quantizeAttributes.js [hunk 23]
```diff
@@ -0,0 +1,281 @@
...
+    if (size === 3) {
+        return createDecodeMatrix3(min, max, range);
+    } else if (size === 4) {
+        return createDecodeMatrix4(min, max, range);
+    } else if (size === 5) {
+        return createDecodeMatrix5(min, max, range);
+    }
+}
+
+function createDecodeMatrix3(min, max, range) {
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Here and below, the intention is for the whole numbers to be floating-point, right?  Append `.0`.

See https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Documentation/Contributors/CodingGuide/README.md#basic-code-construction

File: lib/quantizeAttributes.js (hunk 21 above)
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Can you point me to the specific example of where this much flexibility is needed?

Comment (lasalvavida) [inline lib/quantizeAttributes.js]: See [CesiumMilkTruck-mismatch](https://github.com/lasalvavida/cesium/blob/bf81d91d104fa3d0d1d74ab6a5451070c1f3c75f/Specs/Data/Models/WEB3DQuantizedAttributes/CesiumMilkTruck-Mismatch-Quantized.gltf)

Just the POSITIONS and NORMALS on one primitive in a mesh are quantized, and just the TEXCOORD on a primitive in a different mesh is quantized. 

In practice, I agree, you generally just want to quantize everything or a specific semantic, but I needed those options to help generate potentially breaking test cases.

Comment (pjcozzi) [inline lib/quantizeAttributes.js]: I don't think it is worth it; virtually no user will want to control quantization in such a fine-grained way and the simpler implementation is better for maintenance.  I think it is a better trade off to keep this simple even if it means some breaking cases can't be tested.

**Comment by lasalvavida** [comment]
@pjcozzi Just tests left to do; everything seems to work so far. One question regarding removeUnusedVertices: packBuffers has effectively the same behavior except that it will re-order and un-interleave accessors to optimize buffer size. Should we remove removeUnusedVertices completely?

**Comment by lasalvavida** [comment]
@pjcozzi This is ready for another look.

Also, I don't know why Travis is failing, the tests all pass locally.

Coverage is:
Statements   : 94.06% ( 1949/2072 )
Branches     : 72.38% ( 663/916 )
Functions    : 97.58% ( 121/124 )
Lines        : 94.06% ( 1949/2072 )

vs. master:
Statements   : 93.81% ( 1714/1827 )
Branches     : 73.09% ( 611/836 )
Functions    : 99.05% ( 104/105 )
Lines        : 93.81% ( 1714/1827 )

**Comment by pjcozzi** [comment]
> packBuffers has effectively the same behavior except that it will re-order and un-interleave accessors to optimize buffer size. Should we remove removeUnusedVertices completely?

I'm not completely following; we still need a stage that looks at indices and removes unused vertices.  I don't believe `packBuffers` does that, right?

**Comment by pjcozzi** [comment]
There are CI test failures: https://travis-ci.org/AnalyticalGraphicsInc/gltf-pipeline/builds/135985901

**Comment by pjcozzi** [comment]
> Also, I don't know why Travis is failing, the tests all pass locally.

Ah, just saw this.  It could be a different OS or Node.js version.

File: lib/getBinaryGltf.js (hunk 5 above)
Comment (pjcozzi) [inline lib/getBinaryGltf.js]: @lasalvavida can you check for `process.nextTick()` elsewhere in gltf-pipeline?  I'm pretty sure we should be using it here for consistent Node.js callback async vs. sync behavior.

**Comment by lasalvavida** [comment]
> I'm not completely following; we still need a stage that looks at indices and removes unused vertices. I don't believe packBuffers does that, right?

Yes, sorry I misunderstood. I'll work packBuffers into removeUnusedVertices. That should also fix #57

File: lib/quantizeAttributes.js [hunk 24]
```diff
@@ -0,0 +1,245 @@
...
+ *
+ * @returns glTF with quantized attributes
+ *
+ * @example
+ * var gltf = {
+ *     accessors: {
+ *         accessor_0: {...}
+ *         accessor_1: {...}
+ *     },
+ *     meshes: {
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Is this example old?

File: lib/packBuffers.js [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+
+var byteLengthForComponentType = require('./byteLengthForComponentType');
+var getAccessorByteStride = require('./getAccessorByteStride');
+var numberOfComponentsForType = require('./numberOfComponentsForType');
+
+module.exports = packBuffers;
+
+/**
+ * Repacks the accessed buffer data into contiguous chunks.
+ * Also has the effect of un-interleaving interleaved accessors.
```
Comment (pjcozzi) [inline lib/packBuffers.js]: Is it reasonable/useful to un-interleave as a separate stage since several other stages will also need to depend on it?

File: lib/quantizeAttributes.js (hunk 24 above)
Comment (lasalvavida) [inline lib/quantizeAttributes.js]: It looks okay to me. Do you see something wrong with it?

File: lib/packBuffers.js [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+
+function packGltfBuffer(gltf, bufferId, packBufferViews) {
+    var buffers = gltf.buffers;
+    var buffer = buffers[bufferId];
+    var source = buffer.extras._pipeline.source;
+    var packBuffer = new Uint8Array(source.length);
+    var accessors = getAccessorsByTargetAndByteLength(gltf, bufferId);
+    var offset = 0;
+
+    var targets = [34962, 0, 34963];
```
Comment (pjcozzi) [inline lib/packBuffers.js]: Use `Cesium.WebGLConstants`.  See https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Renderer/WebGLConstants.js

File: lib/packBuffers.js [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+function getAccessorsByTargetAndByteLength(gltf, bufferId) {
+    var accessors = gltf.accessors;
+    var bufferViews = gltf.bufferViews;
+    var accessorsByTargetAndByteLength = {};
+    for (var accessorId in accessors) {
+        var accessor = accessors[accessorId];
+        var bufferView = bufferViews[accessor.bufferView];
+        if (bufferView.buffer === bufferId) {
+            var target = bufferView.target;
+            if (!defined(target)) {
```
Comment (pjcozzi) [inline lib/packBuffers.js]: Use `defaultValue`; it will be more concise.

File: lib/quantizeAttributes.js [hunk 28]
```diff
@@ -0,0 +1,245 @@
...
+        attributeSemantic.indexOf("COLOR") >= 0;
+}
+
+function accessorIsQuantizable(accessor) {
+    // This accessor is already quantized
+    if (accessorIsQuantized(accessor)) {
+        return false;
+    }
+    // Only 32-bit float to 16-bit int quantization is supported
+    return accessor.componentType == 5126;
```
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Same `WebGLConstants` comment here and throughout if needed.

File: lib/packBuffers.js (hunk 25 above)
Comment (lasalvavida) [inline lib/packBuffers.js]: I don't think so; the un-interleaving is just a by-product of packBuffers copying the data into the smallest space possible. 

I could write a separate function to do un-interleaving without removing unused data, but it wouldn't make this one simpler and I can't think of a case where you would want to intentionally keep unused data.

File: lib/quantizeAttributes.js (hunk 24 above)
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Did you see https://github.com/AnalyticalGraphicsInc/gltf-pipeline/pull/56#discussion_r66105451

Perhaps the right way to keep the interface simple and pragmatic and also allow testing the corner cases is to allow the test to pass in an undocumented callback function that takes the right parameters to decide if an attribute should be quantized or not.  That keeps this implementation simple and moves the corner-case complexity (and overall less complexity) to the tests where it is the focus.

File: lib/packBuffers.js (hunk 25 above)
Comment (pjcozzi) [inline lib/packBuffers.js]: OK, perhaps name this `uninterleaveAndPackBuffers` (generally, putting "and" in a name is not ideal, but I think it is valuable here) since other stages will want to depend on it for the uninterleaving part.

File: lib/quantizeAttributes.js (hunk 24 above)
Comment (lasalvavida) [inline lib/quantizeAttributes.js]: I followed the suggestions in that comment and changed the example; that was part of this round of changes. Instead of having the user specify specifically which attributes in which meshes, it is now just options.semantics: [SEMANTIC1, SEMANTIC2 ...].

A callback is an interesting idea, but I'm not sure it's necessary. I needed to generate corner cases for the reader, in the writer it doesn't really matter, so it's probably better just to keep it simple.

Comment (lasalvavida) [inline lib/quantizeAttributes.js]: I'll remove the example gltf hierarchy at the top though. It isn't really necessary anymore since the examples aren't targeting specific id's anymore

Comment (pjcozzi) [inline lib/quantizeAttributes.js]: Ah, right!  This is the glTF, not `options`! 😆

Comment (pjcozzi) [inline lib/quantizeAttributes.js]: OK

**Comment by lasalvavida** [comment]
@pjcozzi I think I hit everything here. Fixing removeUnusedVertices will be part of a separate pull request; it's outside the scope of these changes.

File: bin/gltf-pipeline.js [hunk 29]
```diff
@@ -18,7 +18,8 @@ if (process.argv.length < 3 || defined(argv.h) || defined(argv.help)) {
         '  -i, input=PATH Read unoptimized glTF from the specified file.\n ' +
         '  -o, output=PATH write optimized glTF to the specified file.\n' +
         '  -b, write binary glTF file.\n' +
-        '  -s, writes out separate geometry/animation data files, shader files and textures instead of embedding them in the glTF file.\n  ';
+        '  -s, writes out separate geometry/animation data files, shader files and textures instead of embedding them in the glTF file.\n' + 
+        '  -q, quantize the attributes of this model.\n';
```
Comment (pjcozzi) [inline bin/gltf-pipeline.js]: Also update README.md please.

File: lib/quantizeAttributes.js (hunk 6 above)
Comment (pjcozzi) [inline lib/quantizeAttributes.js]: How hard is it to get test coverage for this file to 100%? https://coveralls.io/builds/6535609/source?filename=lib%2FquantizeAttributes.js

**Comment by pjcozzi** [comment]
Just those trivial comments.  Looking forward to merging this!

**Comment by lasalvavida** [comment]
Updated.

**Comment by pjcozzi** [comment]
Awesome!
